rspace/src/main/scala/coop/rchain/rspace/RSpace.scala

```diff
@@ -165,67 +163,65 @@ class RSpace[F[_], C, P, A, R, K] private[rspace] (
-                                              channelsToIndex
-                                            )
-                                        _ <- logF.debug(
-                                              s"consume: data found for <patterns: $patterns> at <channels: $channels>"
-                                            )
-                                      } yield wrapResult(consumeRef, dataCandidates)
+                             )
+                       _ <- syncF.delay {
+                             eventLog.update(consumeRef +: _)
+                           }
+                       _ <- spanF.mark("event-log-updated")
 
-                                  }
-                         _ <- spanF.mark("extract-consume-candidate")
-                       } yield result
+                       channelToIndexedData <- fetchChannelToIndexData(channels)
+                       _                    <- spanF.mark("channel-to-indexed-data-fetched")
+
+                       options <- extractDataCandidates(
+                                   channels.zip(patterns),
+                                   channelToIndexedData,
+                                   Nil
+                                 ).map(_.sequence)
+                       _ <- spanF.mark("extract-consume-candidate")
+                       result <- options match {
+                                  case None => storeWC(consumeRef)
+                                  case Some(dataCandidates) =>
+                                    for {
+                                      _ <- metricsF.incrementCounter(consumeCommLabel)
+                                      _ <- syncF.delay {
+                                            eventLog.update(
+                                              COMM(
+                                                consumeRef,
+                                                dataCandidates.map(_.datum.source),
+                                                peeks
+                                              ) +: _
+                                            )
+                                          }
+                                      channelsToIndex = channels.zipWithIndex.toMap
+                                      _ <- storePersistentData(
+                                            dataCandidates,
+                                            peeks,
+                                            channelsToIndex
+                                          )
+                                      _ <- logF.debug(
+                                            s"consume: data found for <patterns: $patterns> at <channels: $channels>"
+                                          )
+                                    } yield wrapResult(consumeRef, dataCandidates)
+
+                                }
+                       _ <- spanF.mark("extract-consume-candidate")
+                     } yield result
```

Comment:
this function is big and could use a refactoring. It would be nice now, that you touched it, to clean it up a little bit

Reply:
agreed. Ill try to come up with something